src/service/sync/user-sync-service.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ *  SudoSOS back-end API service.
+ *  Copyright (C) 2024  Study association GEWIS
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as published
+ *  by the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ *  @license
+ */
+
+import User from '../../entity/user/user';
+import WithManager from '../../database/with-manager';
+import { SyncResult, SyncService } from './sync-service';
+import { In } from 'typeorm';
+
+export default class UserSyncService extends WithManager {
+
+  private services: SyncService[];
+
+  constructor(services: SyncService[]) {
+    super();
+    this.services = services;
+  }
+
+  async syncUsers() {
+    const userTypes = this.services.flatMap((s) => s.targets);
+    console.log('Syncing users of types', userTypes);
+
+    const users = await this.manager.find(User, { where: { type: In(userTypes) } });
+    for (const user of users) {
+      const result = await this.sync(user);
+
+      if (result.skipped) {
+        console.log('Sync skipped for user', user.id);
+        continue;
+      }
+
+      if (result.error) {
+        console.error('Sync failed for user', user.id);
+        continue;
+      }
+
+      console.log('Synced user', user.id);
+      if (result.result === false) {
+        console.error('User is detached', user.id);
+        await this.down(user);
+      }
```

Comment:
I think it is better if `sync()` (or `up()`) always returns the sync result or `undefined`/`null` if you have not synchronized anything. If syncing returns an error, I think it is nicer to handle this with a custom exception.

Reply:
Where would you like to catch it? simply in the user-sync-service?

Reply:
Also, I think I prefer an objects with
```
{
  skipped: boolean;
  result: booleanl
}
```

As giving implicit values to undefined | null | boolean does not make the code clearer, and javascript equality checks are wack in general. (falsly vs thruthy)